src/cockpit/samples.py

```diff
@@ -223,18 +235,18 @@ def sample(self, samples):
             self.cgroups_v2 = os.path.exists('/sys/fs/cgroup/cgroup.controllers')
 
         if self.cgroups_v2_path:
-            for path, _, _ in os.walk(self.cgroups_v2_path):
+            for path, _, _, rootfd in os.fwalk(self.cgroups_v2_path):
                 cgroup = path.replace(self.cgroups_v2_path, '')  # TODO: Pathlib?
 
                 if not cgroup:
                     continue
 
-                self.read_cgroup_integer_stat(samples, os.path.join(path, 'memory.current'), cgroup, 'cgroup.memory.usage')
-                self.read_cgroup_integer_stat(samples, os.path.join(path, 'memory.max'), cgroup, 'cgroup.memory.limit')
-                self.read_cgroup_integer_stat(samples, os.path.join(path, 'memory.swap.current'), cgroup, 'cgroup.memory.sw-usage')
-                self.read_cgroup_integer_stat(samples, os.path.join(path, 'memory.swap.max'), cgroup, 'cgroup.memory.sw-limit')
-                self.read_cgroup_integer_stat(samples, os.path.join(path, 'cpu.weight'), cgroup, 'cgroup.cpu.shares')
-                self.read_cgroup_keyed_stat(samples, os.path.join(path, 'cpu.stat'), cgroup, 'cgroup.cpu.usage', 'usage_usec')
+                self.read_cgroup_integer_stat(samples, rootfd, 'memory.current', cgroup, 'cgroup.memory.usage')
```

Comment:
I wonder why not something like

```python
samples['cgroup.memory.usage'][cgroup] = read_cgroup_int(rootfd, 'memory.current')
samples['cgroup.cpu.usage'][cgroup] = read_cgroup_keyed(rootfd, 'cpu.stat', 'usage_usec')
...etc
```

with `read_cgroup_int()` and `read_cgroup_keyed()` being a single functions that open, read, (key search), converts-to-int and returns the result

I wouldn't worry too much about the case of an out-of-bounds value... This is the kernel.  If that happens you can just write in a 0 or just let `int()` throw.

Reply:
Uh, I don't think we want to assign 0 when the file is not found or can't be read, the C code just errors out and doesn't set a sample so I think we want to replicate that behaviour

Reply:
None?